// botbuilder-ai/lib/qnaMakerRecognizer.js
"use strict";
/**
 * @module botbuilder-dialogs-adaptive
 */
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const adaptive_expressions_1 = require("adaptive-expressions");
const botbuilder_dialogs_1 = require("botbuilder-dialogs");
const qnaMaker_1 = require("./qnaMaker");
const qnamaker_interfaces_1 = require("./qnamaker-interfaces");
const intentPrefix = 'intent=';
/**
 * A recognizer which uses QnAMaker KB to recognize intents.
 */
class QnAMakerRecognizer extends botbuilder_dialogs_1.Recognizer {
    /**
     * Initializes a new instance of `QnAMakerRecognizer`.
     * @param hostname Hostname of QnAMaker KB.
     * @param knowledgeBaseId Id of QnAMaker KB.
     * @param endpointKey Endpoint key of QnAMaker KB.
     */
    constructor(hostname, knowledgeBaseId, endpointKey) {
        super();
        /**
         * Number of results you want.
         */
        this.top = new adaptive_expressions_1.IntExpression(3);
        /**
         * Threshold for the results.
         */
        this.threshold = new adaptive_expressions_1.NumberExpression(0.3);
        /**
         * Desired RankerType.
         */
        this.rankerType = new adaptive_expressions_1.StringExpression(qnamaker_interfaces_1.RankerTypes.default);
        /**
         * Whether to include the dialog name metadata for QnA context.
         */
        this.includeDialogNameInMetadata = new adaptive_expressions_1.BoolExpression(true);
        /**
         * An expression to evaluate to set QnAId parameter.
         */
        this.qnaId = new adaptive_expressions_1.IntExpression(0);
        if (hostname) {
            this.hostname = new adaptive_expressions_1.StringExpression(hostname);
        }
        if (knowledgeBaseId) {
            this.knowledgeBaseId = new adaptive_expressions_1.StringExpression(knowledgeBaseId);
        }
        if (endpointKey) {
            this.endpointKey = new adaptive_expressions_1.StringExpression(endpointKey);
        }
    }
    getConverter(property) {
        switch (property) {
            case 'knowledgeBaseId':
                return new adaptive_expressions_1.StringExpressionConverter();
            case 'hostname':
                return new adaptive_expressions_1.StringExpressionConverter();
            case 'endpointKey':
                return new adaptive_expressions_1.StringExpressionConverter();
            case 'top':
                return new adaptive_expressions_1.IntExpressionConverter();
            case 'threshold':
                return new adaptive_expressions_1.NumberExpressionConverter();
            case 'rankerType':
                return new adaptive_expressions_1.StringExpressionConverter();
            case 'includeDialogNameInMetadata':
                return new adaptive_expressions_1.BoolExpressionConverter();
            case 'metadata':
                return new adaptive_expressions_1.ArrayExpressionConverter();
            case 'context':
                return new adaptive_expressions_1.ObjectExpressionConverter();
            case 'qnaId':
                return new adaptive_expressions_1.IntExpressionConverter();
            default:
                return super.getConverter(property);
        }
    }
    /**
     * Gets results of the call to QnA maker KB.
     * @param dc Context object containing information for a single turn of coversation with a user.
     * @param activity The incoming activity received from the user. The text value is used as the query to QnA Maker.
     * @param telemetryProperties Additional properties to be logged to telemetry.
     * @param telemetryMetrics Additional metrics to be logged to telemetry.
     */
    recognize(dc, activity, telemetryProperties, telemetryMetrics) {
        return __awaiter(this, void 0, void 0, function* () {
            // identify matched intents
            const recognizerResult = {
                text: activity.text,
                intents: {},
                entities: {},
            };
            if (!activity.text) {
                recognizerResult.intents['None'] = { score: 1 };
                return recognizerResult;
            }
            const filters = [];
            if (this.includeDialogNameInMetadata && this.includeDialogNameInMetadata.getValue(dc.state)) {
                const metadata = {
                    name: 'dialogName',
                    value: dc.activeDialog && dc.activeDialog.id,
                };
                filters.push(metadata);
            }
            // if there is $qna.metadata set add to filters
            const externalMetadata = this.metadata && this.metadata.getValue(dc.state);
            if (externalMetadata) {
                filters.push(...externalMetadata);
            }
            // calling QnAMaker to get response
            const qnaMaker = this.getQnAMaker(dc);
            const qnaMakerOptions = {
                context: this.context && this.context.getValue(dc.state),
                scoreThreshold: this.threshold && this.threshold.getValue(dc.state),
                strictFilters: filters,
                top: this.top && this.top.getValue(dc.state),
                qnaId: this.qnaId && this.qnaId.getValue(dc.state),
                rankerType: this.rankerType && this.rankerType.getValue(dc.state),
                isTest: this.isTest,
            };
            const answers = yield qnaMaker.getAnswers(dc.context, qnaMakerOptions);
            if (answers && answers.length > 0) {
                let topAnswer;
                for (let i = 0; i < answers.length; i++) {
                    const answer = answers[i];
                    if (!topAnswer || answer.score > topAnswer.score) {
                        topAnswer = answer;
                    }
                }
                if (topAnswer.answer.trim().toLowerCase().startsWith(intentPrefix)) {
                    recognizerResult.intents[topAnswer.answer.trim().substr(intentPrefix.length).trim()] = {
                        score: topAnswer.score,
                    };
                }
                else {
                    recognizerResult.intents[QnAMakerRecognizer.qnaMatchIntent] = { score: topAnswer.score };
                }
                recognizerResult.entities['answer'] = [topAnswer.answer];
                recognizerResult.entities['$instance'] = {
                    answer: [
                        Object.assign(topAnswer, {
                            startIndex: 0,
                            endIndex: activity.text.length,
                        }),
                    ],
                };
                recognizerResult['answers'] = answers;
            }
            else {
                recognizerResult.intents['None'] = { score: 1 };
            }
            this.trackRecognizerResult(dc, 'QnAMakerRecognizerResult', this.fillRecognizerResultTelemetryProperties(recognizerResult, telemetryProperties), telemetryMetrics);
            return recognizerResult;
        });
    }
    /**
     * Gets an instance of `QnAMaker`.
     * @param dc The dialog context used to access state.
     */
    getQnAMaker(dc) {
        const endpointKey = this.endpointKey && this.endpointKey.getValue(dc.state);
        const hostname = this.hostname && this.hostname.getValue(dc.state);
        const knowledgeBaseId = this.knowledgeBaseId && this.knowledgeBaseId.getValue(dc.state);
        const endpoint = {
            endpointKey: endpointKey,
            host: hostname,
            knowledgeBaseId: knowledgeBaseId,
        };
        return new qnaMaker_1.QnAMaker(endpoint);
    }
}
QnAMakerRecognizer.$kind = 'Microsoft.QnAMakerRecognizer';
QnAMakerRecognizer.qnaMatchIntent = 'QnAMatch';
exports.QnAMakerRecognizer = QnAMakerRecognizer;
