// botbuilder-testing/lib/dialogTestClient.js
"use strict";
/**
 * @module botbuilder-testing
 */
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
const botbuilder_core_1 = require("botbuilder-core");
const botbuilder_dialogs_1 = require("botbuilder-dialogs");
/**
 * A client for testing dialogs in isolation.
 */
class DialogTestClient {
    /**
     * Creates a [DialogTestClient](xref:botbuilder-testing.DialogTestClient) to test a [Dialog](xref:botbuilder-dialogs.Dialog) without having to create a full-fledged adapter.
     * @param channelOrAdapter The `channelId` or the [TestAdapter](xref:botbuilder-core.TestAdapter) to be used for the test.
     * @param targetDialog The [Dialog](xref:botbuilder-dialogs.Dialog) to be tested. This will be the root dialog for the test client.
     * @param initialDialogOptions Optional. Additional argument(s) to pass to the [Dialog](xref:botbuilder-dialogs.Dialog) being started.
     * @param middlewares Optional. A [Middleware](xref:botbuilder-core.Middleware) list to be added to the test adapter.
     * @param conversationState Optional. A [ConversationState](xref:botbuilder-core.ConversationState) instance to use in the test client.
     */
    constructor(channelOrAdapter, targetDialog, initialDialogOptions, middlewares, conversationState) {
        this._dialogContext = null;
        this.conversationState = conversationState || new botbuilder_core_1.ConversationState(new botbuilder_core_1.MemoryStorage());
        const dialogState = this.conversationState.createProperty('DialogState');
        this._callback = this.getDefaultCallback(targetDialog, initialDialogOptions || null, dialogState);
        if (typeof channelOrAdapter == 'string') {
            const channelIdToUse = channelOrAdapter;
            this._testAdapter = new botbuilder_core_1.TestAdapter(this._callback, { channelId: channelIdToUse }).use(new botbuilder_core_1.AutoSaveStateMiddleware(this.conversationState));
        }
        else {
            const testAdapterToUse = channelOrAdapter;
            this._testAdapter = testAdapterToUse;
        }
        this.addUserMiddlewares(middlewares);
    }
    /**
     * Gets a reference for the DialogContext.
     * @remarks
     * This property will be null until at least one activity is sent to DialogTestClient.
     */
    get dialogContext() {
        return this._dialogContext;
    }
    /**
     * Send an activity into the dialog.
     * @returns a TestFlow that can be used to assert replies etc
     * @param activity an activity potentially with text
     *
     * ```javascript
     * DialogTest.send('hello').assertReply('hello yourself').then(done);
     * ```
     */
    sendActivity(activity) {
        return __awaiter(this, void 0, void 0, function* () {
            yield this._testAdapter.receiveActivity(activity);
            return this._testAdapter.activityBuffer.shift();
        });
    }
    /**
     * Get the next reply waiting to be delivered (if one exists)
     */
    getNextReply() {
        return this._testAdapter.activityBuffer.shift();
    }
    /**
     * @private
     */
    getDefaultCallback(targetDialog, initialDialogOptions, dialogState) {
        return (turnContext) => __awaiter(this, void 0, void 0, function* () {
            const dialogSet = new botbuilder_dialogs_1.DialogSet(dialogState);
            dialogSet.add(targetDialog);
            this._dialogContext = yield dialogSet.createContext(turnContext);
            this.dialogTurnResult = yield this._dialogContext.continueDialog();
            if (this.dialogTurnResult.status === botbuilder_dialogs_1.DialogTurnStatus.empty) {
                this.dialogTurnResult = yield this._dialogContext.beginDialog(targetDialog.id, initialDialogOptions);
            }
        });
    }
    /**
     * @private
     */
    addUserMiddlewares(middlewares) {
        if (middlewares != null) {
            middlewares.forEach((middleware) => {
                this._testAdapter.use(middleware);
            });
        }
    }
}
exports.DialogTestClient = DialogTestClient;
