// adaptive-expressions/lib/datetimeFormatConverter.js
"use strict";
/* eslint-disable @typescript-eslint/no-unused-vars */
/**
 * @module adaptive-expressions
 */
/**
 * Copyright (c) Microsoft Corporation. All rights reserved.
 * Licensed under the MIT License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
var State;
(function (State) {
    State[State["None"] = 0] = "None";
    State[State["LowerD1"] = 1] = "LowerD1";
    State[State["LowerD2"] = 2] = "LowerD2";
    State[State["LowerD3"] = 3] = "LowerD3";
    State[State["LowerD4"] = 4] = "LowerD4";
    State[State["LowerF1"] = 5] = "LowerF1";
    State[State["LowerF2"] = 6] = "LowerF2";
    State[State["LowerF3"] = 7] = "LowerF3";
    State[State["LowerF4"] = 8] = "LowerF4";
    State[State["LowerF5"] = 9] = "LowerF5";
    State[State["LowerF6"] = 10] = "LowerF6";
    State[State["LowerF7"] = 11] = "LowerF7";
    State[State["CapitalF1"] = 12] = "CapitalF1";
    State[State["CapitalF2"] = 13] = "CapitalF2";
    State[State["CapitalF3"] = 14] = "CapitalF3";
    State[State["CapitalF4"] = 15] = "CapitalF4";
    State[State["CapitalF5"] = 16] = "CapitalF5";
    State[State["CapitalF6"] = 17] = "CapitalF6";
    State[State["CapitalF7"] = 18] = "CapitalF7";
    State[State["LowerG"] = 19] = "LowerG";
    State[State["LowerH1"] = 20] = "LowerH1";
    State[State["LowerH2"] = 21] = "LowerH2";
    State[State["CapitalH1"] = 22] = "CapitalH1";
    State[State["CapitalH2"] = 23] = "CapitalH2";
    State[State["CapitalK"] = 24] = "CapitalK";
    State[State["LowerM1"] = 25] = "LowerM1";
    State[State["LowerM2"] = 26] = "LowerM2";
    State[State["CapitalM1"] = 27] = "CapitalM1";
    State[State["CapitalM2"] = 28] = "CapitalM2";
    State[State["CapitalM3"] = 29] = "CapitalM3";
    State[State["CapitalM4"] = 30] = "CapitalM4";
    State[State["LowerS1"] = 31] = "LowerS1";
    State[State["LowerS2"] = 32] = "LowerS2";
    State[State["LowerT1"] = 33] = "LowerT1";
    State[State["LowerT2"] = 34] = "LowerT2";
    State[State["LowerY1"] = 35] = "LowerY1";
    State[State["LowerY2"] = 36] = "LowerY2";
    State[State["LowerY3"] = 37] = "LowerY3";
    State[State["LowerY4"] = 38] = "LowerY4";
    State[State["LowerY5"] = 39] = "LowerY5";
    State[State["LowerZ1"] = 40] = "LowerZ1";
    State[State["LowerZ2"] = 41] = "LowerZ2";
    State[State["LowerZ3"] = 42] = "LowerZ3";
    State[State["InSingleQuoteLiteral"] = 43] = "InSingleQuoteLiteral";
    State[State["InDoubleQuoteLiteral"] = 44] = "InDoubleQuoteLiteral";
    State[State["EscapeSequence"] = 45] = "EscapeSequence";
})(State || (State = {}));
/**
 * Convert a CSharp style datetime format string to a Moment.js style datetime format string. Ref: https://docs.microsoft.com/en-us/dotnet/standard/base-types/custom-date-and-time-format-strings
 * @param fmtString A CSharp style datetime format string. Ref: https://devhints.io/moment
 * @returns A Momengt.js style datetime format string.
 */
function convertCSharpDateTimeToMomentJS(fmtString) {
    let fmtResult = '';
    let fmtState = State.None;
    let lTokenBuffer = '';
    if (fmtString.length === 0) {
        return fmtResult;
    }
    if (fmtString.length === 1) {
        switch (fmtString) {
            case 'R':
            case 'r':
                throw Error(`RFC 1123 not supported  in MomentJS`);
            case 'O':
            case 'o':
                fmtString = 'YYYY-MM-DDTHH:mm:ss.SSSSSSSZ';
                break;
            case 'U':
                throw new Error(`Universal Fulll Format not supported in MomentJS`);
            case 'u':
                throw new Error(`Universal Sortable Format not supported in MomentJS`);
        }
    }
    const changeState = (newState) => {
        switch (fmtState) {
            case State.LowerD1:
                fmtResult += 'D';
                break;
            case State.LowerD2:
                fmtResult += 'DD';
                break;
            case State.LowerD3:
                fmtResult += 'ddd';
                break;
            case State.LowerD4:
                fmtResult += 'dddd';
                break;
            case State.LowerF1:
            case State.CapitalF1:
                fmtResult += 'S';
                break;
            case State.LowerF2:
            case State.CapitalF2:
                fmtResult += 'SS';
                break;
            case State.LowerF3:
            case State.CapitalF3:
                fmtResult += 'SSS';
                break;
            case State.LowerF4:
            case State.CapitalF4:
                fmtResult += 'SSSS';
                break;
            case State.LowerF5:
            case State.CapitalF5:
                fmtResult += 'SSSSS';
                break;
            case State.LowerF6:
            case State.CapitalF6:
                fmtResult += 'SSSSSS';
                break;
            case State.LowerF7:
            case State.CapitalF7:
                fmtResult += 'SSSSSSS';
                break;
            case State.LowerG:
                throw Error('Era not supported in MomentJS');
            case State.LowerH1:
                fmtResult += 'h';
                break;
            case State.LowerH2:
                fmtResult += 'hh';
                break;
            case State.CapitalH1:
                fmtResult += 'H';
                break;
            case State.CapitalH2:
                fmtResult += 'HH';
                break;
            case State.LowerM1:
                fmtResult += 'm';
                break;
            case State.LowerM2:
                fmtResult += 'mm';
                break;
            case State.CapitalM1:
                fmtResult += 'M';
                break;
            case State.CapitalM2:
                fmtResult += 'MM';
                break;
            case State.CapitalM3:
                fmtResult += 'MMM';
                break;
            case State.CapitalM4:
                fmtResult += 'MMMM';
                break;
            case State.LowerS1:
                fmtResult += 's';
                break;
            case State.LowerS2:
                fmtResult += 'ss';
                break;
            case State.LowerT1:
                fmtResult += 'A';
                break;
            case State.LowerT2:
                fmtResult += 'A';
                break;
            case State.LowerY1:
                fmtResult += 'YY';
                break;
            case State.LowerY2:
                fmtResult += 'YY';
                break;
            case State.LowerY3:
                fmtResult += 'YYYY';
                break;
            case State.LowerY4:
                fmtResult += 'YYYY';
                break;
            case State.LowerY5:
                fmtResult += 'Y';
                break;
            case State.LowerZ1:
            case State.LowerZ2:
                fmtResult += 'ZZ';
                break;
            case State.LowerZ3:
                fmtResult += 'Z';
                break;
            case State.InSingleQuoteLiteral:
            case State.InDoubleQuoteLiteral:
            case State.EscapeSequence:
                for (const lCharacter of lTokenBuffer) {
                    fmtResult += lCharacter;
                }
                break;
        }
        lTokenBuffer = '';
        fmtState = newState;
    };
    for (const character of fmtString) {
        if (fmtState === State.EscapeSequence) {
            lTokenBuffer += character;
            changeState(State.None);
        }
        else if (fmtState === State.InDoubleQuoteLiteral) {
            if (character == '`') {
                changeState(State.None);
            }
            else {
                lTokenBuffer += character;
            }
        }
        else if (fmtState == State.InSingleQuoteLiteral) {
            if (character == "'") {
                changeState(State.None);
            }
            else {
                lTokenBuffer += character;
            }
        }
        else {
            switch (character) {
                case 'd':
                    switch (fmtState) {
                        case State.LowerD1:
                            fmtState = State.LowerD2;
                            break;
                        case State.LowerD2:
                            fmtState = State.LowerD3;
                            break;
                        case State.LowerD3:
                            fmtState = State.LowerD4;
                            break;
                        case State.LowerD4:
                            break;
                        default:
                            changeState(State.LowerD1);
                            break;
                    }
                    break;
                case 'f':
                    switch (fmtState) {
                        case State.LowerF1:
                            fmtState = State.LowerF2;
                            break;
                        case State.LowerF2:
                            fmtState = State.LowerF3;
                            break;
                        case State.LowerF3:
                            fmtState = State.LowerF4;
                            break;
                        case State.LowerF4:
                            fmtState = State.LowerF5;
                            break;
                        case State.LowerF5:
                            fmtState = State.LowerF6;
                            break;
                        case State.LowerF6:
                            fmtState = State.LowerF7;
                            break;
                        case State.LowerF7:
                            break;
                        default:
                            changeState(State.LowerF1);
                            break;
                    }
                    break;
                case 'F':
                    switch (fmtState) {
                        case State.CapitalF1:
                            fmtState = State.CapitalF2;
                            break;
                        case State.CapitalF2:
                            fmtState = State.CapitalF3;
                            break;
                        case State.CapitalF3:
                            fmtState = State.CapitalF4;
                            break;
                        case State.CapitalF4:
                            fmtState = State.CapitalF5;
                            break;
                        case State.CapitalF5:
                            fmtState = State.CapitalF6;
                            break;
                        case State.CapitalF6:
                            fmtState = State.CapitalF7;
                            break;
                        case State.CapitalF7:
                            break;
                        default:
                            changeState(State.CapitalF1);
                            break;
                    }
                    break;
                case 'g':
                    switch (fmtState) {
                        case State.LowerG:
                            break;
                        default:
                            changeState(State.LowerG);
                            break;
                    }
                    break;
                case 'h':
                    switch (fmtState) {
                        case State.LowerH1:
                            fmtState = State.LowerH2;
                            break;
                        case State.LowerH2:
                            break;
                        default:
                            changeState(State.LowerH1);
                            break;
                    }
                    break;
                case 'H':
                    switch (fmtState) {
                        case State.CapitalH1:
                            fmtState = State.CapitalH2;
                            break;
                        case State.CapitalH2:
                            break;
                        default:
                            changeState(State.CapitalH1);
                            break;
                    }
                    break;
                case 'K':
                    changeState(State.None);
                    fmtResult += 'Z';
                    break;
                case 'm':
                    switch (fmtState) {
                        case State.LowerM1:
                            fmtState = State.LowerM2;
                            break;
                        case State.LowerM2:
                            break;
                        default:
                            changeState(State.LowerM1);
                            break;
                    }
                    break;
                case 'M':
                    switch (fmtState) {
                        case State.CapitalM1:
                            fmtState = State.CapitalM2;
                            break;
                        case State.CapitalM2:
                            fmtState = State.CapitalM3;
                            break;
                        case State.CapitalM3:
                            fmtState = State.CapitalM4;
                            break;
                        case State.CapitalM4:
                            break;
                        default:
                            changeState(State.CapitalM1);
                            break;
                    }
                    break;
                case 's':
                    switch (fmtState) {
                        case State.LowerS1:
                            fmtState = State.LowerS2;
                            break;
                        case State.LowerS2:
                            break;
                        default:
                            changeState(State.LowerS1);
                            break;
                    }
                    break;
                case 't':
                    switch (fmtState) {
                        case State.LowerT1:
                            fmtState = State.LowerT2;
                            break;
                        case State.LowerT2:
                            break;
                        default:
                            changeState(State.LowerT1);
                            break;
                    }
                    break;
                case 'y':
                    switch (fmtState) {
                        case State.LowerY1:
                            fmtState = State.LowerY2;
                            break;
                        case State.LowerY2:
                            fmtState = State.LowerY3;
                            break;
                        case State.LowerY3:
                            fmtState = State.LowerY4;
                            break;
                        case State.LowerY4:
                            fmtState = State.LowerY5;
                            break;
                        case State.LowerY5:
                            break;
                        default:
                            changeState(State.LowerY1);
                            break;
                    }
                    break;
                case 'z':
                    switch (fmtState) {
                        case State.LowerZ1:
                            fmtState = State.LowerZ2;
                            break;
                        case State.LowerZ2:
                            fmtState = State.LowerZ3;
                            break;
                        case State.LowerZ3:
                            break;
                        default:
                            changeState(State.LowerZ1);
                            break;
                    }
                    break;
                case ':':
                    changeState(State.None);
                    fmtResult += ':';
                    break;
                case '/':
                    changeState(State.None);
                    fmtResult += '/';
                    break;
                case '`':
                    changeState(State.InDoubleQuoteLiteral);
                    break;
                case "'":
                    changeState(State.InSingleQuoteLiteral);
                    break;
                case '%':
                    changeState(State.None);
                    break;
                case '\\':
                    changeState(State.EscapeSequence);
                    break;
                default:
                    changeState(State.None);
                    fmtResult += character;
                    break;
            }
        }
    }
    if (fmtState == State.EscapeSequence ||
        fmtState == State.InDoubleQuoteLiteral ||
        fmtState == State.InSingleQuoteLiteral) {
        throw Error(`Invalid Format String`);
    }
    changeState(State.None);
    return fmtResult;
}
exports.convertCSharpDateTimeToMomentJS = convertCSharpDateTimeToMomentJS;
